File: nanopay/src/net/nanopay/cico/paymentCard/PaymentCardTransientDAO.java
package net.nanopay.cico.paymentCard;

import net.nanopay.cico.paymentCard.model.*;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.core.FObject;
import foam.core.X;

public class PaymentCardTransientDAO 
  extends ProxyDAO
{
  public PaymentCardTransientDAO(X x, DAO delegate) {
    setX(x);
    setDelegate(delegate);
  }

  @Override
  public FObject put_(X x, FObject obj) {
    PaymentCard card = (PaymentCard) obj;
    card.setCvv(null);
    return getDelegate().put_(x, obj);
  }
}

File: nanopay/src/net/nanopay/cico/paymentCard/PaymentCardTransientDAO.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.cico.paymentCard',
  name: 'PaymentCardTransientDAO',
  extends: 'foam.dao.ProxyDAO',

  javaImports: [
    'net.nanopay.cico.paymentCard.model.*',
    'foam.dao.DAO',
    'foam.core.FObject',
    'foam.core.X'
  ],

  axioms: [
    {
      name: 'javaExtras',
      buildJavaClass: function(cls) {
        cls.extras.push(`
          public PaymentCardTransientDAO(X x, DAO delegate) {
            setX(x);
            setDelegate(delegate);
          }    
        `
        );
      }
    }
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        PaymentCard card = (PaymentCard) obj;
        card.setCvv(null);
        return getDelegate().put_(x, obj);
      `
    }
  ]
});
